libvast/src/taxonomies.cpp

```diff
@@ -21,18 +21,43 @@
 
 #include <algorithm>
 #include <deque>
-#include <stack>
+#include <string_view>
 
 namespace vast {
 
 bool operator==(const taxonomies& lhs, const taxonomies& rhs) {
   return lhs.concepts == rhs.concepts && lhs.models == rhs.models;
 }
 
+static bool
+contains(const std::map<std::string, type_set>& seen, const std::string& x,
+         relational_operator op, const vast::data& data) {
+  std::string::size_type pos = 0;
+  while ((pos = x.find('.', pos)) != std::string::npos) {
+    auto i = seen.find(std::string{std::string_view{x.c_str(), pos}});
+    if (i != seen.end()) {
+      // A prefix of x matches an existing layout.
+      auto field = x.substr(pos + 1);
+      return std::any_of(i->second.value.begin(), i->second.value.end(),
+                         [&](const type& t) {
+                           if (auto r = caf::get_if<record_type>(&t)) {
+                             if (auto f = r->find(field))
+                               return compatible(f->type, op, data);
+                           }
+                           return false;
+                         });
```

Comment:
It would a bit more readable if you factored the predicate into a named lambda.

Reply:
I tried and it doesn't. The formatting is the same bar indentation, but we get an additional indirection, so I'd rather keep it like that.